pkg/flow/streaming/flow_sliding_window.go

```diff
@@ -0,0 +1,334 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package streaming
+
+import (
+	"container/heap"
+	"context"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/apache/skywalking-banyandb/pkg/flow"
+)
+
+type TriggerResult bool
+
+const (
+	FIRE     TriggerResult = true
+	CONTINUE               = false
+)
+
+var (
+	_ flow.Operator       = (*TumblingTimeWindows)(nil)
+	_ flow.WindowAssigner = (*TumblingTimeWindows)(nil)
+	_ flow.Window         = (*timeWindow)(nil)
+
+	_ TriggerContext = (*triggerContext)(nil)
+)
+
+func (f *streamingFlow) Window(w flow.WindowAssigner) flow.WindowedFlow {
+	switch v := w.(type) {
+	case *TumblingTimeWindows:
+		v.errorHandler = f.drainErr
+		f.ops = append(f.ops, v)
+	default:
+		f.drainErr(errors.New("window type is not supported"))
+	}
+
+	return &windowedFlow{
+		f:  f,
+		wa: w,
+	}
+}
+
+func (s *windowedFlow) AllowedLateness(lateness time.Duration) flow.Flow {
+	switch v := s.wa.(type) {
+	case *TumblingTimeWindows:
+		v.lateness = lateness.Milliseconds()
+	default:
+		s.f.drainErr(errors.New("lateness is not supported"))
+	}
+	return s.f
+}
+
+type TumblingTimeWindows struct {
+	// internal state of the sliding time window
+	flow.ComponentState
+	// errorHandler is the error handler and set by the streamingFlow
+	errorHandler func(error)
+	// For TumblingTimeWindows
+	// Unit: Millisecond
+	size int64
+	// lateness is the maximum allowed lateness for incoming elements
+	// Unit: Millisecond
+	lateness int64
+
+	// guard snapshots
+	snapshotsMu sync.Mutex
+	snapshots   map[int64]flow.AggregationOp
+	acc         flow.AggregationOp
+
+	currentWatermark int64
+	// guard timerHeap
+	timerMu   sync.Mutex
+	timerHeap *flow.DedupPriorityQueue
+
+	// aggregationFactory is the factory for creating aggregation operator
+	aggregationFactory flow.AggregationOpFactory
+
+	// For api.Operator
+	in           chan flow.StreamRecord
+	out          chan flow.StreamRecord
+	done         chan struct{}
+	purgedWindow chan timeWindow
+}
+
+func (s *TumblingTimeWindows) In() chan<- flow.StreamRecord {
+	return s.in
+}
+
+func (s *TumblingTimeWindows) Out() <-chan flow.StreamRecord {
+	return s.out
+}
+
+func (s *TumblingTimeWindows) Setup(ctx context.Context) error {
+	if s.acc == nil {
+		s.acc = s.aggregationFactory()
+	}
+	// start processing
+	s.Add(1)
+	go s.receive()
+	// start emitting
+	s.Add(1)
+	go s.emit()
+
+	return nil
+}
+
+func (s *TumblingTimeWindows) emit() {
+	defer s.Done()
+	for w := range s.purgedWindow {
+		if err := func(window timeWindow) error {
+			s.snapshotsMu.Lock()
+			defer s.snapshotsMu.Unlock()
```

Comment:
Locking a channel writing is not good practice.

Reply:
Fixed